File: src/main/java/com/dongtronic/diabot/listener/ConversionListener.java
package com.dongtronic.diabot.listener;

import com.dongtronic.diabot.converters.BloodGlucoseConverter;
import com.dongtronic.diabot.converters.GlucoseUnit;
import com.dongtronic.diabot.data.ConversionDTO;
import com.dongtronic.diabot.exceptions.UnknownUnitException;
import net.dv8tion.jda.core.entities.*;
import net.dv8tion.jda.core.events.message.guild.GuildMessageReceivedEvent;
import net.dv8tion.jda.core.hooks.ListenerAdapter;
import net.dv8tion.jda.core.managers.GuildController;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class ConversionListener extends ListenerAdapter {
  private Logger logger = LoggerFactory.getLogger(ConversionListener.class);

  private static Pattern inlinePattern = Pattern.compile("^.*_([0-9]{1,3}\\.?[0-9]?)_.*$");
  private static Pattern separatePattern = Pattern.compile("^([0-9]{1,3}\\.?[0-9]?)$");

  @Override
  public void onGuildMessageReceived(GuildMessageReceivedEvent event) {
    if (event.getAuthor().isBot()) return;

    TextChannel channel = event.getChannel();

    Message message = event.getMessage();
    String messageText = message.getContentRaw();

    Matcher inlineMatcher = inlinePattern.matcher(messageText);
    Matcher separateMatcher = separatePattern.matcher(messageText);

    String number = "";

    if(inlineMatcher.matches()) {
      number = inlineMatcher.group(1);
    } else if (separateMatcher.matches()) {
      number = separateMatcher.group(1);
    }

    if(number.length() == 0) {
      return;
    }

    try {
      ConversionDTO result = BloodGlucoseConverter.convert(number, null);

      if (result.getInputUnit() == GlucoseUnit.MMOL) {
        channel.sendMessage(String.format("%s mmol/L is %s mg/dL", result.getOriginal(), result.getConverted())).queue();
      } else if (result.getInputUnit() == GlucoseUnit.MGDL) {
        channel.sendMessage(String.format("%s mg/dL is %s mmol/L", result.getOriginal(), result.getConverted())).queue();
      } else {
        String reply = String.join(
            "%n",
            "*I'm not sure if you gave me mmol/L or mg/dL, so I'll give you both.*",
            "%s mg/dL is **%s mmol/L**",
            "%s mmol/L is **%s mg/dL**");

        channel.sendMessage(String.format(reply, result.getOriginal(), result.getMmol(), result.getOriginal(),
            result.getMgdl())).queue();
      }

    } catch (IllegalArgumentException ex) {
      // Ignored on purpose
      logger.warn("IllegalArgumentException occurred but was ignored in BG conversion");
    } catch (UnknownUnitException ex) {
      // Ignored on purpose
    }

  }

}

File: src/main/java/com/dongtronic/diabot/listener/ConversionListener.kt
package com.dongtronic.diabot.listener

import com.dongtronic.diabot.converters.BloodGlucoseConverter
import com.dongtronic.diabot.converters.GlucoseUnit
import com.dongtronic.diabot.data.ConversionDTO
import com.dongtronic.diabot.exceptions.UnknownUnitException
import com.dongtronic.diabot.util.Patterns
import net.dv8tion.jda.core.events.message.guild.GuildMessageReceivedEvent
import net.dv8tion.jda.core.hooks.ListenerAdapter
import org.slf4j.LoggerFactory

class ConversionListener : ListenerAdapter() {
    private val logger = LoggerFactory.getLogger(ConversionListener::class.java)

    override fun onGuildMessageReceived(event: GuildMessageReceivedEvent) {
        if (event.author.isBot) return

        val channel = event.channel

        val message = event.message
        val messageText = message.contentRaw

        val inlineMatcher = Patterns.inlineBgPattern.matcher(messageText)
        val separateMatcher = Patterns.separateBgPattern.matcher(messageText)
        val unitMatcher = Patterns.unitBgPattern.matcher(messageText)

        var numberString = ""
        var unitString = ""

        if (unitMatcher.matches()) {
            numberString = unitMatcher.group(4)
            unitString = unitMatcher.group(5)
        }

        if (inlineMatcher.matches()) {
            numberString = inlineMatcher.group(1)
        } else if (separateMatcher.matches()) {
            numberString = separateMatcher.group(1)
        }

        if (numberString.isEmpty()) {
            return
        }

        try {
            val result: ConversionDTO? = if (unitString.length > 1) {
                BloodGlucoseConverter.convert(numberString, unitString)
            } else {
                BloodGlucoseConverter.convert(numberString, null)
            }

            when {
                result!!.inputUnit === GlucoseUnit.MMOL -> channel.sendMessage(String.format("%s mmol/L is %s mg/dL", result!!.mmol, result.mgdl)).queue()
                result!!.inputUnit === GlucoseUnit.MGDL -> channel.sendMessage(String.format("%s mg/dL is %s mmol/L", result!!.mgdl, result.mmol)).queue()
                else -> {
                    val reply = arrayOf(
                            "*I'm not sure if you gave me mmol/L or mg/dL, so I'll give you both.*",
                            "%s mg/dL is **%s mmol/L**",
                            "%s mmol/L is **%s mg/dL**").joinToString(
                            "%n")

                    channel.sendMessage(String.format(reply, numberString, result!!.mmol, numberString,
                            result.mgdl)).queue()
                }
            }

            // #20: Reply with :smirk: when value is 69 mg/dL or 6.9 mmol/L
            if (numberString == "6.9" || numberString == "69") {
                event.message.addReaction("\uD83D\uDE0F").queue()
            }

            // #36: Reply with :100: when value is 100 mg/dL or 5.5 mmol/L
            if (numberString == "5.5" || numberString == "100") {
                event.message.addReaction("\uD83D\uDCAF").queue()
            }

        } catch (ex: IllegalArgumentException) {
            // Ignored on purpose
            logger.warn("IllegalArgumentException occurred but was ignored in BG conversion")
        } catch (ex: UnknownUnitException) {
            // Ignored on purpose
        }

    }

}
